src/java.base/share/classes/javax/crypto/KEM.java

```diff
@@ -0,0 +1,613 @@
+                }
+            }
+            if (iape != null) throw iape;
+            if (ike != null) throw ike;
+            if (nsae != null) {
+                throw new InvalidKeyException("No installed provider found", nsae);
+            }
+            throw new InvalidKeyException("No installed provider supports this key: "
+                    + sk.getClass().getName(), re);
+        }
+    }
+
+    // If delayed provider selection is needed
+    private final DelayedKEM delayed;
+
+    // otherwise
+    private final KEMSpi spi;
+    private final Provider provider;
+
+    private KEM(KEMSpi spi, Provider provider) {
+        assert spi != null;
+        assert provider != null;
+        this.delayed = null;
+        this.spi = spi;
+        this.provider = provider;
+    }
+
+    private KEM(DelayedKEM delayed) {
+        assert delayed != null;
+        this.delayed = delayed;
+        this.spi = null;
+        this.provider = null;
+    }
+
+    /**
+     * Returns a {@code KEM} object that implements the specified algorithm.
+     *
+     * @param algorithm the name of the KEM algorithm.
+     *          See the {@code KEM} section in the <a href=
+     *          "{@docRoot}/../specs/security/standard-names.html#kem-algorithms">
+     *          Java Security Standard Algorithm Names Specification</a>
+     *          for information about standard KEM algorithm names.
+     * @return the new {@code KEM} object
+     * @throws NoSuchAlgorithmException if no {@code Provider} supports a
+     *         {@code KEM} implementation for the specified algorithm
+     * @throws NullPointerException if the input argument is {@code null}
+     */
+    public static KEM getInstance(String algorithm)
+            throws NoSuchAlgorithmException {
+        List<Provider.Service> list = GetInstance.getServices(
+                "KEM",
+                Objects.requireNonNull(algorithm, "null algorithm name"));
+        if (list.isEmpty()) {
+            throw new NoSuchAlgorithmException(algorithm + " KEM not available");
+        }
+        return new KEM(new DelayedKEM(list.toArray(new Provider.Service[0])));
+    }
+
+    /**
+     * Returns a {@code KEM} object that implements the specified algorithm
+     * and supports the specified {@code KEMParameterSpec} parameters
+     * from the specified security provider.
+     *
+     * @param algorithm the name of the KEM algorithm.
+     *          See the {@code KEM} section in the <a href=
+     *          "{@docRoot}/../specs/security/standard-names.html#kem-algorithms">
+     *          Java Security Standard Algorithm Names Specification</a>
+     *          for information about standard KEM algorithm names.
+     * @param provider the provider. If {@code null}, this method is equivalent
+     *                 to {@link #getInstance(String)}.
+     * @return the new {@code KEM} object
+     * @throws NoSuchAlgorithmException if no {@code Provider} supports a
```

Comment:
This sounds like the provider argument is ignored. Maybe add some more words to take into account both cases "if a {@code provider} is specified and it does not support the specified KEM algorithm, or if {@code provider} is {@code null} and there is no provider that supports a KEM implementation of the  specified algorithm"